builder/docker/communicator.go

```diff
@@ -60,40 +60,83 @@ func (c *Communicator) Start(remote *packer.RemoteCmd) error {
 	return nil
 }
 
+// Upload uploads a file to the docker container
 func (c *Communicator) Upload(dst string, src io.Reader, fi *os.FileInfo) error {
+	if fi == nil {
+		return c.uploadReader(dst, src)
+	}
+	return c.uploadFile(dst, src, fi)
+}
+
+// uploadReader writes an io.Reader to a temporary file before uploading
+func (c *Communicator) uploadReader(dst string, src io.Reader) error {
 	// Create a temporary file to store the upload
 	tempfile, err := ioutil.TempFile(c.HostDir, "upload")
-	if err != nil {
-		return err
-	}
 	defer os.Remove(tempfile.Name())
+	defer tempfile.Close()
 
-	// Copy the contents to the temporary file
 	_, err = io.Copy(tempfile, src)
 	if err != nil {
+		return fmt.Errorf("Failed to copy upload file to tempfile: %s", err)
+	}
+	tempfile.Seek(0, 0)
+	fi, err := tempfile.Stat()
+	if err != nil {
+		return fmt.Errorf("Error getting tempfile info: %s", err)
+	}
+	return c.uploadFile(dst, tempfile, &fi)
+}
+
+// uploadFile uses docker cp to copy the file from the host to the container
+func (c *Communicator) uploadFile(dst string, src io.Reader, fi *os.FileInfo) error {
+
+	// command format: docker cp /path/to/infile containerid:/path/to/outfile
+	log.Printf("Copying to %s on container %s.", dst, c.ContainerId)
+
+	localCmd := exec.Command("docker", "cp", "-",
+		fmt.Sprintf("%s:%s", c.ContainerId, filepath.Dir(dst)))
+
+	stderrP, err := localCmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("Failed to open pipe: %s", err)
+	}
+
+	stdin, err := localCmd.StdinPipe()
+	if err != nil {
+		return fmt.Errorf("Failed to open pipe: %s", err)
+	}
+
+	if err := localCmd.Start(); err != nil {
 		return err
 	}
 
-	if fi != nil {
-		tempfile.Chmod((*fi).Mode())
+	archive := tar.NewWriter(stdin)
```

Comment:
I'm not immediately clear on why we need to create an archive with the new way of doing things.  Haven't we already copied at this point using docker cp?